src/tasks/LargeFileUploadTask.ts

```diff
@@ -0,0 +1,344 @@
+  }
+
+  /**
+   * Extracts the upload session information from a parsable object.
+   *
+   * @param {Parsable} parsable - The parsable object containing the upload session information.
+   * @returns {UploadSession} - The extracted upload session information.
+   */
+  private extractSessionInfo(parsable: Parsable): UploadSession {
+    const { expirationDateTime, nextExpectedRanges, odataType, uploadUrl } = parsable as Partial<UploadSession>;
+    if (!nextExpectedRanges || !uploadUrl || nextExpectedRanges.length === 0) {
+      throw new Error("Upload session is invalid");
+    }
+    return {
+      expirationDateTime: expirationDateTime ?? null,
+      nextExpectedRanges: nextExpectedRanges ?? null,
+      odataType: odataType ?? null,
+      uploadUrl: uploadUrl ?? null,
+    };
+  }
+
+  /**
+   * Generates a list of upload slice requests based on the remaining ranges to be uploaded.
+   *
+   * @private
+   * @returns {UploadSlice<T>[]} An array of UploadSlice objects representing the upload requests.
+   */
+  private getUploadSliceRequests(): UploadSlice<T>[] {
+    const uploadSlices: UploadSlice<T>[] = [];
+    const rangesRemaining = this.rangesRemaining;
+    const session = this.Session;
+    rangesRemaining.forEach(range => {
+      let currentRangeBegin = range[0];
+      while (currentRangeBegin <= range[1]) {
+        const nextSliceSize = this.nextSliceSize(currentRangeBegin, range[1]);
+        const uploadRequest = new UploadSlice<T>(
+          this.requestAdapter,
+          session.uploadUrl!,
+          currentRangeBegin,
+          currentRangeBegin + nextSliceSize - 1,
+          range[1] + 1,
+          this.parsableFactory,
+          this.errorMappings,
+          this.seekableStreamReader,
+        );
+        uploadSlices.push(uploadRequest);
+        currentRangeBegin += nextSliceSize;
+      }
+    });
+    return uploadSlices;
+  }
+
+  /**
+   * Calculates the size of the next slice to be uploaded.
+   *
+   * @param {number} currentRangeBegin - The beginning of the current range.
+   * @param {number} currentRangeEnd - The end of the current range.
+   * @returns {number} - The size of the next slice.
+   */
+  private nextSliceSize(currentRangeBegin: number, currentRangeEnd: number): number {
+    const sizeBasedOnRange = currentRangeEnd - currentRangeBegin + 1;
+    return sizeBasedOnRange > this.maxSliceSize ? this.maxSliceSize : sizeBasedOnRange;
+  }
+
+  /**
+   * @private
+   * Parses the upload session response and returns a nested number array of ranges pending upload
+   * @param uploadSession
+   */
+  private getRangesRemaining(uploadSession: UploadSession): number[][] {
+    // nextExpectedRanges: https://dev.onedrive.com/items/upload_large_files.htm
+    // Sample: ["12345-55232","77829-99375"]
+    // Also, second number in range can be blank, which means 'until the end'
+    const ranges: number[][] = [];
+    uploadSession.nextExpectedRanges?.forEach(rangeString => {
```

Comment:
instead of foreach + push use map. You'll avoid growing the array multiple times.